src/items.rs

```diff
@@ -757,9 +757,12 @@ pub fn format_impl(
             result.push_str(&outer_indent_str);
         }
 
-        if result.ends_with('{') && !context.config.empty_item_single_line() {
+        if result.ends_with('{')
+            && (!context.config.empty_item_single_line() || trimmed_last_line_width(&result) == 1)
```

Comment:
I think you can reuse the variable `need_newline` https://github.com/rust-lang-nursery/rustfmt/pull/3050/files#diff-5db152a52bdaeae9dacd35f43a4a78ddR722 which is true if a newline before `{` was needed.

Reply:
Thanks for the tip! True. It's possible to reuse need_newline here, and that made the code much simpler.